st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+CustomValidator = extend_with_default(CustomValidator)
+
+
+class NotFoundException(Exception):
+    pass
+
+
+class Response(webob.Response):
+    def __init__(self, body=None, status=None, headerlist=None, app_iter=None, content_type=None,
+                 *args, **kwargs):
+        # Do some sanity checking, and turn json_body into an actual body
+        if app_iter is None and body is None and ('json_body' in kwargs or 'json' in kwargs):
+            if 'json_body' in kwargs:
+                json_body = kwargs.pop('json_body')
+            else:
+                json_body = kwargs.pop('json')
+            body = json_encode(json_body).encode('UTF-8')
+
+            if content_type is None:
+                content_type = 'application/json'
+
+        super(Response, self).__init__(body, status, headerlist, app_iter, content_type,
+                                       *args, **kwargs)
+
+    def _json_body__set(self, value):
+        self.body = json_encode(value).encode('UTF-8')
+
+
+class Router(object):
+    def __init__(self, arguments=None, debug=False, auth=True):
+        self.debug = debug
+        self.auth = auth
+
+        self.arguments = arguments or {}
+
+        self.spec = {}
+        self.spec_resolver = None
+        self.routes = routes.Mapper()
+
+    def add_spec(self, spec, transforms):
+        info = spec.get('info', {})
+        LOG.debug('Adding API: %s %s', info.get('title', 'untitled'), info.get('version', '0.0.0'))
+
+        self.spec = spec
+        self.spec_resolver = validate_spec(copy.deepcopy(self.spec))
+
+        for filter in transforms:
+            for (path, methods) in six.iteritems(spec['paths']):
+                if not re.search(filter, path):
+                    continue
+
+                for (method, endpoint) in six.iteritems(methods):
+                    conditions = {
+                        'method': [method.upper()]
+                    }
+
+                    connect_kw = {}
+                    if 'x-requirements' in endpoint:
+                        connect_kw['requirements'] = endpoint['x-requirements']
+
+                    m = self.routes.submapper(_api_path=path, _api_method=method,
+                                              conditions=conditions)
+                    for transform in transforms[filter]:
+                        m.connect(None, re.sub(filter, transform, path), **connect_kw)
+
+        for route in sorted(self.routes.matchlist, key=lambda r: r.routepath):
+            LOG.debug('Route registered: %+6s %s', route.conditions['method'][0], route.routepath)
+
+    def match(self, req):
+        path = req.path
+
+        if len(path) > 1 and path.endswith('/'):
+            path = path[:-1]
+
+        match = self.routes.match(path, req.environ)
+
+        if match is None:
+            raise NotFoundException('No route matches "%s" path' % req.path)
+
+        # To account for situation when match may return multiple values
```

Comment:
Hm wouldn't multiple values indicate a programmer error - aka the route is not set up correctly since there should only be one controller match for each path?